docs/design/features/greenthreads.md

```diff
@@ -0,0 +1,231 @@
+
+##  Implementation constraints and details
+As part of the experiment, we developed an implementation of green threads in the runtimelab repo. See dotnet/runtimelab at feature/green-threads (github.com) and Commits · dotnet/aspnetcore (github.com)
+The general outcome was a version of green threads that uses the segmented stack model, and provides a small number of new low level apis to work with green threads, and a fairly small number of higher level apis were changed to take advantage of green threads.
+### Fundamental api changes
+-	Thread.IsGreenThread – Used to identify if the current thread is a green thread
+-	Task.RunAsGreenThread(Action) – Similar to Task.Run, but runs code on a green thread, using the green thread threadpool.
+-	Task.Wait() – When called on a green thread, will yield the green thread until the task waited on is complete, then will schedule the green thread to continue
+### Higher level api changes
+-	Socket and file apis were changed to support green threads by transitioning synchronous apis to async apis when called on a green thread.
+-	Kestrel in ASP.NET was modified to support writing a response via a green thread.
+
+### Implementation constraints discovered during research
+A general principle of the green thread effort was that normal, already existing .NET code should be able to work when run on a green thread. There are a number of specific features of .NET which complicate implementation of green threads, and this document will attempt to describe the .NET features which are known to be problematic, and why.
+1. Support for byref parameters
+  *	Known to be problematic, as they cannot simply be copied to the GC heap, and can currently only be reported to the GC as part of the stack walking logic
+2. Support for conversion of byrefs to pointers
+  * Certain byrefs, if they are known to point to the stack can safely be converted into pointers in the product today
+  * As pointers are not trackable by the runtime (and this is not fixable, as it is legal to export pointers to a non-process visible location, and then bring them back), it is not possible to safely implement a scheme which allows stacks to be moved during GC in all circumstances
+  * This complicates the design of a green thread system, as it cannot use a contiguous stack for green threads, as it would need to be prohibitively large
+  * Other systems that have successfully transitioned to green threads such as Go/Java do not have this capability, and are able to implement green thread approach which do not have the Hot-Split problem
+3.	Support for function pointers
```

Comment:
```suggestion
  - Known to be problematic, as they cannot simply be copied to the GC heap, and can currently only be reported to the GC as part of the stack walking logic
2. Support for conversion of byrefs to pointers
  - Certain byrefs, if they are known to point to the stack can safely be converted into pointers in the product today
  - As pointers are not trackable by the runtime (and this is not fixable, as it is legal to export pointers to a non-process visible location, and then bring them back), it is not possible to safely implement a scheme which allows stacks to be moved during GC in all circumstances
  - This complicates the design of a green thread system, as it cannot use a contiguous stack for green threads, as it would need to be prohibitively large
  - Other systems that have successfully transitioned to green threads such as Go/Java do not have this capability, and are able to implement green thread approach which do not have the Hot-Split problem
3. Support for function pointers
```